### PR TITLE
188649391 modal refresh bug

### PR DESCRIPTION
**Pull request by knil-maloon**
### *External Module new alert:
Screenshot 2024-12-05 at 8 32 42 AM

### *Logging:
Screenshot 2024-12-05 at 8 31 25 AM

## Overview
New message on Hub Updates if the Mapper project detects that some of the projects don't have the module activated or a configuration setting created.

## This PR Updates:
1.  Hub Updates index.php, adding a new alert to notify us that there are projects that don't have the module activated and/or a configuration setting created. This prevents the hub from working normally.
2. New class checkIfModuleIsEnabledOnProjects on ProjectData.php where this check and update takes place.
3. Minor conditional bug fix on HarmonistHubExternalModule.php on redcap_module_link_check_display.
4. Reformated code on all the files.

### CODE REVIEW

File: hub-updates/enable_module_and_settings_AJAX.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+include_once(__DIR__ . "/../projects.php");
+include_once(__DIR__ . "/../classes/HubUpdates.php");
+
+$pid = $_REQUEST['pid'];
+
+ProjectData::checkIfModuleIsEnabledOnProjects($module, $pidsArray, $pid, true);
+$message = "&message=E";
+
+header("location:" . $module->getUrl('hub-updates/index.php') . $message);
```
Comment (paulj3) [inline hub-updates/enable_module_and_settings_AJAX.php:13]: Is this file actually used in an ajax request?

Comment (knil-maloon) [reply hub-updates/enable_module_and_settings_AJAX.php:13]: Not an ajax call but a form call upon clicking the enable button. That file is to print message and refresh

**Review by paulj3 — APPROVED** [review]
Great work on this Eva! I suspect you must have worked pretty hard on this, given that you just uncovered the issue yesterday. I think this is a good fix.

File: hub-updates/enable_module_and_settings_AJAX.php (hunk 1 above)
Comment (paulj3) [reply hub-updates/enable_module_and_settings_AJAX.php:13]: okay, i see. I recommend taking AJAX off the file name then.

File: hub-updates/index.php [hunk 2]
```diff
@@ -1,386 +1,524 @@
...
+                } else {
+                    if (array_key_exists('message', $_REQUEST) && ($_REQUEST['message'] == 'T')) {
+                        $message = "<strong>" . ProjectData::HUB_SURVEY_THEME_NAME . "</strong> has been successfully updated. Check the logs to see which surveys have been updated.";
+                    } else {
+                        if (array_key_exists('message', $_REQUEST) && ($_REQUEST['message'] == 'V')) {
+                            $message = "The surveys have been succesfully created. Check the logs to see which instruments have been updated.";
+                        } else {
+                            if (array_key_exists('message', $_REQUEST) && ($_REQUEST['message'] == 'E')) {
+                                $message = "The module and settings have been enabled on the projects.";
+                            }
```
[the remark this answers is not in the document]
Comment (knil-maloon) [reply hub-updates/index.php]: That is a great idea and it makes it cleaner! Just pushed the change.

**Comment by knil-maloon** [comment]
Thanks! I wanted to have this running soon so we can test well the Hub tomorrow :)

File: hub-updates/index.php [hunk 3]
```diff
@@ -230,41 +230,20 @@ function changeFormUrlPDF(id) {
...
-    if (array_key_exists('message', $_REQUEST)) { ?>
+    $message = [
+        'S' => "The Data Dictionary has been successfully updated.",
+        'R' => "The variables have been successfully <strong>added</strong> to the resolved list.",
+        'U' => "The variables have been successfully <strong>removed</strong> from the resolved list.",
+        'L' => "The Data Dictionary has been successfully updated.",
+        'T' => "<strong>" . ProjectData::HUB_SURVEY_THEME_NAME . "</strong> has been successfully updated. Check the logs to see which surveys have been updated.",
+        'V' => "The surveys have been succesfully created. Check the logs to see which instruments have been updated.",
+        'E' => "The module and settings have been enabled on the projects.",
+    ];
```
Comment (paulj3) [inline hub-updates/index.php:241]: This is great! But I think two small things could really make it better. I think this array should live as a static array on a class.  Then you should use this static array to set these letters in the jacascript.  then use the array again here to turn the letters into messages.

Comment (knil-maloon) [reply hub-updates/index.php]: For now I created a new Messages class and I'm calling a function with the messages on it. The current call does not use ajax but directly the form call so I would have to do more changes to adapt and I think we can do these later if that's ok :)
For now I think the new class will be useful to store all messages :)

Comment (knil-maloon) [reply hub-updates/index.php]: let me know what you think about the new changes and I'll merge

File: hub-updates/index.php [hunk 4]
```diff
@@ -1,386 +1,497 @@
...
-                    }else{
-                        success_message = "U";
+                    if (option == 'save') {
+                        success_message = "S";
+                    } else {
+                        success_message = "R";
                     }
+                } else {
+                    success_message = "U";
+                }
```
[the remark this answers is not in the document]
Comment (knil-maloon) [reply hub-updates/index.php:135]: I think this needs more thinking. I am not sure I follow when you say use the keys as we still need these letters on the url when refreshing.

[the remark this answers is not in the document]
Comment (knil-maloon) [reply hub-updates/index.php:135]: Went ahead and merged to be able to test. We can talk about this after :)

**Pull request merged**